### DISCUSSION TITLE
Congestion control

### DISCUSSION THREAD

**Discussion by jkarthic** [paullouisageneau/libdatachannel#1101] [Q&A] [answered]
I have successfully used this repo to stream my live video packets via webrtc, and view it in a browser natively. Now the next step for me is to adapt my video encoder's bitrate based on the estimated network bandwidth between the streamer and the client. How can I achieve this via libdatachannel? Does it provide the estimated network bandwidth via some APIs? Or do I need to do integrate some other third party library for congestion control implementation? I am totally clueless. Could someone please guide me in the right direction?
Congestion control is such a crucial part of WebRTC streaming. I just want to know how people using libdatachannel generally handle this part.

**Comment by paullouisageneau** [comment, marked as answer]
libdatachannel currently does not have an integrated congestion control algorithm for media transport, but there is an ongoing work in progress from [vagonhq](https://github.com/vagonhq/libdatachannel/tree/twcc-new-handler) to implement Google Congestion Control.

> How can I achieve this via libdatachannel? Does it provide the estimated network bandwidth via some APIs?

Note that congestion control algorithms typically do not work this way. Instead, they provide a target bitrate for the encoder. The algorithm slowly increases the bitrate until the network shows signs of congestion, then it decreases the bitrate and repeats.

> Congestion control is such a crucial part of WebRTC streaming. I just want to know how people using libdatachannel generally handle this part.

This kind of congestion control is only possible if you have direct control on the encoder target bitrate. For instance people often forward streams from external cameras on which you have no control. Also in practice you don't need congestion control for audio only, nor for low-bitrate video.

**Reply by jkarthic** [reply]
Thanks for the clarification. Now I understand better the usecases better.
In our use-case, we would like to do some high resolution, high quality streaming. Hence we need to stream at higher bitrates, where a congestion control algo is required for it to work at scale.
